python/lsst/ctrl/bps/generic_workflow.py

```diff
@@ -581,3 +638,102 @@ def validate(self):
         """
         # Make sure a directed acyclic graph
         assert nx.algorithms.dag.is_directed_acyclic_graph(self)
+
+    def add_workflow_source(self, workflow):
+        """Add given workflow as new source to this workflow.
+
+        Parameters
+        ----------
+        workflow : `lsst.ctrl.bps.GenericWorkflow`
+        """
+        # Find source nodes in self.
+        self_sources = [n for n in self if self.in_degree(n) == 0]
+        _LOG.debug("self_sources = %s", self_sources)
+
+        # Find sink nodes of workflow.
+        new_sinks = [n for n in workflow if workflow.out_degree(n) == 0]
+        _LOG.debug("new sinks = %s", new_sinks)
+
+        # Add new workflow nodes to self graph and make new edges.
+        self.add_nodes_from(workflow.nodes(data=True))
+        self.add_edges_from(workflow.edges())
+        for source in self_sources:
+            for sink in new_sinks:
+                self.add_edge(sink, source)
+
+        # Files are stored separately so copy them.
+        for job_name in workflow:
+            self.add_job_inputs(job_name, workflow.get_job_inputs(job_name, data=True))
+            self.add_job_outputs(job_name, workflow.get_job_outputs(job_name, data=True))
+            self.add_executable(workflow.get_job(job_name).executable)
+
+    def add_final(self, final):
+        """Add special final job/workflow to the generic workflow.
+
+        Parameters
+        ----------
+        final : `lsst.ctrl.bps.GenericWorkflowJob` or
```

Comment:
You may want to end this line with `\` (line continuation). If I recall correctly, either `sphinx` or `package-docs` wasn't entirely happy with a type specification spanning over more than one line.